packages/adapter-utils/src/utils.ts

```diff
@@ -101,33 +125,112 @@ const fieldMapperGenerator = (
 }
-export type TransformFunc = (args: TransformFuncArgs) => Promise<Value> | Value | undefined
 
-export type TransformFuncSync = (args: TransformFuncArgs) => lowerDashTypes.NonPromise<Value> | undefined
+const recurseIntoValue = (
+  newVal: Value,
+  transformFunc: (value: Value, keyPathID?: ElemID, field?: Field) => Value,
+  strict: boolean,
+  allowEmpty: boolean,
+  isAsync: boolean,
+  keyPathID?: ElemID,
+  field?: Field,
+  fieldType?: TypeElement,
+): Value => {
+  if (newVal === undefined) {
+    return undefined
+  }
 
-type TransformValuesBaseArgs = {
-  values: Value
-  type: ObjectType | TypeMap | MapType | ListType
-  strict?: boolean
-  pathID?: ElemID
-  isTopLevel?: boolean
-  allowEmpty?: boolean
+  if (isReferenceExpression(newVal)) {
+    return newVal
+  }
+
+  const listMapFunc = isAsync
+    ? (list: Iterable<Value>,
+      mapFunc: (t: Value, index: number) => Promise<unknown>) => toArrayAsync(mapAsync(list, mapFunc))
+    : _.map
+  const objMapFunc = isAsync
+    ? (obj: Iterable<Value>, mapFunc: (val: Value, key: string) => Promise<unknown>) => mapValuesAsync(obj, mapFunc)
+    : _.mapValues
+
+  if (field && isListType(fieldType)) {
+    const transformListInnerValue = (item: Value, index?: number): Value => (
+      transformFunc(
+        item,
+        !_.isUndefined(index) ? keyPathID?.createNestedID(String(index)) : keyPathID,
+        new Field(
+          field.parent,
+          field.name,
+          fieldType.refInnerType,
+          field.annotations
+        ),
+      )
+    )
+    if (!_.isArray(newVal)) {
+      if (strict) {
+        log.debug(`Array value and isListType mis-match for field - ${field.name}. Got non-array for ListType.`)
+      }
+      return transformListInnerValue(newVal)
+    }
+    return listMapFunc(newVal, (v: Value, i: number) => transformListInnerValue(v, i))
+  }
+  if (_.isArray(newVal)) {
+    // Even fields that are not defined as ListType can have array values
+    return listMapFunc(
+      newVal, (item: Value, index: number) => transformFunc(item, keyPathID?.createNestedID(String(index)), field)
+    )
+  }
+
+  if (isObjectType(fieldType) || isMapType(fieldType)) {
+    if (!_.isPlainObject(newVal)) {
+      if (strict) {
+        log.debug(`Value mis-match for field ${field?.name} - value is not an object`)
+      }
+      // _.isEmpty returns true for primitive values (boolean, number)
+      // but we do not want to omit those, we only want to omit empty
+      // objects, arrays and strings. we don't need to check for objects here
+      // because we cannot get here with an object
+      const valueIsEmpty = (Array.isArray(newVal) || _.isString(newVal)) && _.isEmpty(newVal)
+      return (valueIsEmpty && !allowEmpty) ? undefined : newVal
+    }
+    const fieldMapper = fieldMapperGenerator(fieldType, newVal)
+    return objMapFunc(
+      newVal,
+      (value: Value, key: string) => transformFunc(value, keyPathID?.createNestedID(key), fieldMapper(key))
```

Comment:
similar note, types can be inferred
```suggestion
      (value, key) => transformFunc(value, keyPathID?.createNestedID(key), fieldMapper(key))
```

Reply:
same here, getting build failures

Reply:
newer TS versions would be able to handle it but I guess our old one doesn't...
never mind then